File: backend/src/main/java/com/fptgang/backend/model/Contract.java
package com.fptgang.backend.model;

import jakarta.annotation.Nullable;
import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

import java.math.BigDecimal;
import java.time.LocalDateTime;

@Entity
@Data
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Contract {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long contractId;

    @OneToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "project_id", nullable = false, unique = true)
    private Project project;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "freelancer_id", nullable = false)
    private Account freelancer;

    @OneToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "proposal_id", nullable = false, unique = true)
    private Proposal proposal;

    @Column(nullable = false, precision = 10, scale = 2)
    private BigDecimal budget;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private ContractStatus status;

    @Column
    @Nullable
    private LocalDateTime signedAt;

    @Column
    @Nullable
    private LocalDateTime terminatedAt;

    public enum ContractStatus {
        UNSIGNED,
        SIGNED,
        TERMINATED
    }

    @OneToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "file_id", unique = true)
    private File contractFile;

    @CreationTimestamp
    private LocalDateTime createdAt;

    @UpdateTimestamp
    private LocalDateTime updatedAt;
}

File: gendb/src/engine/model/Contract.ts
import {Account} from "./Account";
import {Project} from "./Project";

export enum ContractStatus {
  SIGNED = 'SIGNED',
  TERMINATED = 'TERMINATED',
  UNSIGNED = 'UNSIGNED'
}

export class Contract {
  contract_id: number | null;
  budget: number;
  created_at: Date;
  status: ContractStatus;
  updated_at: Date | null;
  file_id: number | null;
  freelancer_id: number;
  project_id: number;
  proposal_id: number;
  signed_at: Date | null;
  terminated_at: Date | null;

  freelancer: Account;
  project: Project;

  constructor(data: Partial<Contract> = {}) {
    this.contract_id = data.contract_id ?? null;
    this.budget = data.budget ?? 0;
    this.created_at = data.created_at ?? new Date();
    this.status = data.status ?? ContractStatus.UNSIGNED;
    this.updated_at = data.updated_at ?? null;
    this.file_id = data.file_id ?? null;
    this.freelancer_id = data.freelancer_id ?? 0;
    this.project_id = data.project_id ?? 0;
    this.proposal_id = data.proposal_id ?? 0;
    this.freelancer = data.freelancer ?? {} as Account;
    this.project = data.project ?? {} as Project;
    this.signed_at = data.signed_at ?? null;
    this.terminated_at = data.terminated_at ?? null;
  }

  static dump(contracts: Contract[]): string {
    if (!contracts.length) return "";

    const columns = [
      "contract_id",
      "budget",
      "created_at",
      "status",
      "updated_at",
      "signed_at",
      "terminated_at",
      "file_id",
      "freelancer_id",
      "project_id",
      "proposal_id"
    ];

    let sql = `INSERT INTO hirable.contract (${columns.join(", ")}) VALUES\n`;

    const values = contracts.map(contract => {
      const contractValues = [
        contract.contract_id !== null ? contract.contract_id : 'NULL',
        contract.budget.toFixed(2),
        contract.created_at ? `'${contract.created_at.toISOString().slice(0, 19).replace('T', ' ')}'` : 'NULL',
        `'${contract.status}'`,
        contract.updated_at ? `'${contract.updated_at.toISOString().slice(0, 19).replace('T', ' ')}'` : 'NULL',
        contract.signed_at ? `'${contract.signed_at.toISOString().slice(0, 19).replace('T', ' ')}'` : 'NULL',
        contract.terminated_at ? `'${contract.terminated_at.toISOString().slice(0, 19).replace('T', ' ')}'` : 'NULL',
        contract.file_id !== null ? contract.file_id : 'NULL',
        contract.freelancer_id,
        contract.project_id,
        contract.proposal_id
      ];

      return `(${contractValues.join(", ")})`;
    });

    sql += values.join(",\n");
    sql += ";";

    return sql;
  }
}

export default Contract;